pkg/router/template/router.go

```diff
@@ -1007,12 +1007,16 @@ func (r *templateRouter) removeRouteInternal(route *routev1.Route) {
 // numberOfEndpoints returns the number of endpoints
 // Must be called while holding r.lock
 func (r *templateRouter) numberOfEndpoints(id ServiceUnitKey) int32 {
-	var eps = 0
+	var count = 0
 	svc, ok := r.findMatchingServiceUnit(id)
-	if ok && len(svc.EndpointTable) > eps {
-		eps = len(svc.EndpointTable)
+
+	// Count endpoints that are near identical (same ID)
+	// as the same endpoint to properly calculate endpoint weights
+	endpoints := filterDuplicateEndpointsByID(svc.EndpointTable)
+	if ok && len(endpoints) > count {
```

Comment:
If `findMatchingServiceUnit` does not find a match, it returns the zero value, which has empty `EndpointTable`.  This means that checking `ok` is redundant.  Furthermore, after this change, you are using `svc.EndpointTable` before checking `ok`, which looks wrong (if the check _were_ needed—for example, if `svc` were a pointer value—then `svc.EndpointTable` would be a nil-pointer dereference).  How about doing away with `ok` entirely?